### PR TITLE
Refactor metrics related code and fix calculation for step time metrics

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
# Description

* This PR removes `writer` from the list of values returned by `setup_mesh_and_model`. This simiplifies `setup_mesh_and_model` as it is unnecessary to create a TensorBoard summary writer object in that method. 
* This PR also refactors the metrics related code in all trainers (train.py, sft_trainer.py, grpo_trainer.py, elastic_train.py).
* This PR also fixes the calculation for metric `perf/step_time_seconds` for training. Currently, this metric for step `x` includes the time taken for checkpointing + HLO dump + eval + profiling by step `x-1`. This PR fixes this by ensuring that time taken by checkpointing + HLO dump + eval + profiling by step `x` is added to metric `perf/step_time_seconds` for step `x` only. Depends on profiling fixes [#1833](https://github.com/AI-Hypercomputer/maxtext/pull/1833).

Thanks to @richjames0 for contributing!

# Tests
E2E testing with train.py and sft_trainer.py - verified the metrics are correctly uploaded to TensorBoard

### Logs for refactored code with metric fix:
```
steps=10 profiler=xplane checkpoint_period=7

# training + checkpointing
completed step: 0, seconds: 27.544, TFLOP/s/device: 1.534, Tokens/s/device: 37.177, total_weights: 3573, loss: 0.993

# training + profiling activated
completed step: 1, seconds: 1.105, TFLOP/s/device: 38.242, Tokens/s/device: 927.006, total_weights: 3253, loss: 0.993

# training
completed step: 2, seconds: 0.996, TFLOP/s/device: 42.403, Tokens/s/device: 1027.874, total_weights: 1692, loss: 0.912
completed step: 3, seconds: 0.996, TFLOP/s/device: 42.426, Tokens/s/device: 1028.421, total_weights: 2944, loss: 1.105
completed step: 4, seconds: 0.994, TFLOP/s/device: 42.517, Tokens/s/device: 1030.628, total_weights: 3466, loss: 0.948

# training + profiling deactivated
completed step: 5, seconds: 16.416, TFLOP/s/device: 2.573, Tokens/s/device: 62.376, total_weights: 2420, loss: 0.947

# training
completed step: 6, seconds: 0.982, TFLOP/s/device: 43.037, Tokens/s/device: 1043.238, total_weights: 1655, loss: 1.059

# training + checkpointing
completed step: 7, seconds: 35.911, TFLOP/s/device: 1.176, Tokens/s/device: 28.515, total_weights: 3335, loss: 1.000

# training
completed step: 8, seconds: 0.981, TFLOP/s/device: 43.063, Tokens/s/device: 1043.866, total_weights: 1676, loss: 0.851
completed step: 9, seconds: 0.986, TFLOP/s/device: 42.833, Tokens/s/device: 1038.301, total_weights: 1676, loss: 0.965
```

### Logs for current codebase:
```
steps=10 profiler=xplane checkpoint_period=7

# training
completed step: 0, seconds: 18.461, TFLOP/s/device: 2.288, Tokens/s/device: 55.468, total_weights: 3573, loss: 0.993

# training + checkpointing at step=0 + profiling activated at step=1
completed step: 1, seconds: 9.213, TFLOP/s/device: 4.585, Tokens/s/device: 111.152, total_weights: 3253, loss: 0.993

# training
completed step: 2, seconds: 0.155, TFLOP/s/device: 271.907, Tokens/s/device: 6591.186, total_weights: 1692, loss: 0.912
completed step: 3, seconds: 0.844, TFLOP/s/device: 50.029, Tokens/s/device: 1212.730, total_weights: 2944, loss: 1.105
completed step: 4, seconds: 0.969, TFLOP/s/device: 43.616, Tokens/s/device: 1057.288, total_weights: 3466, loss: 0.948
completed step: 5, seconds: 0.970, TFLOP/s/device: 43.543, Tokens/s/device: 1055.509, total_weights: 2420, loss: 0.947

# training + profiling deactivated at step=5
completed step: 6, seconds: 16.443, TFLOP/s/device: 2.569, Tokens/s/device: 62.277, total_weights: 1655, loss: 1.059

# training
completed step: 7, seconds: 0.015, TFLOP/s/device: 2803.873, Tokens/s/device: 67967.609, total_weights: 3335, loss: 1.000

# training + checkpointing at step=7
completed step: 8, seconds: 34.001, TFLOP/s/device: 1.242, Tokens/s/device: 30.117, total_weights: 1676, loss: 0.851

# training
completed step: 9, seconds: 0.013, TFLOP/s/device: 3224.914, Tokens/s/device: 78173.906, total_weights: 1676, loss: 0.965

```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/metric_logger.py [hunk 1]
```diff
@@ -98,23 +112,22 @@ def write_metrics_locally(self, metrics, step):
...
 
       gcs_filename = os.path.join(self.config.metrics_dir, metrics_filename)
       max_logging.log(f"Moving file {metrics_filename} to GCS...")
       gcs_utils.upload_blob(gcs_filename, metrics_filename)
       max_logging.log(f"File {metrics_filename} moved successfully!")
-      running_metrics = []  # reset running_metrics to empty list
-    return running_metrics
+      self.running_gcs_metrics = []  # reset running_metrics to empty list
 
   def write_metrics_to_tensorboard(self, metrics, step, is_training):
```
Comment (A9isha) [inline MaxText/metric_logger.py:118]: This is not used anywhere I think

File: MaxText/train.py [hunk 2]
```diff
@@ -775,15 +755,10 @@ def train_loop(config, recorder, state=None):
...
-  num_model_parameters = max_utils.calculate_num_params_from_pytree(state.params)
-  max_logging.log(f"number parameters: {num_model_parameters/1e9:.3f} billion")
-  per_device_tflops, _, _ = maxtext_utils.calculate_tflops_training_per_device(config)
-  per_device_tokens = maxtext_utils.calculate_tokens_training_per_device(config)
+  metric_logger = MetricLogger(config=config, learning_rate_schedule=learning_rate_schedule)
 
   # Write train config params, num model params, and XLA flags to tensorboard
-  max_utils.add_text_to_summary_writer("num_model_parameters", str(num_model_parameters), writer)
-  max_utils.add_text_to_summary_writer("libtpu_init_args", os.environ["LIBTPU_INIT_ARGS"], writer)
-  maxtext_utils.add_config_to_summary_writer(config, writer)
```
Comment (A9isha) [inline MaxText/train.py:786]: In order to verify that the logs are identical to the version prior to refactoring, were you able to take a write out the logs and then diff?

File: MaxText/train.py [hunk 3]
```diff
@@ -161,7 +146,7 @@ def save_checkpoint(
       )
 
   # specify chunk_byte_size to force orbax to control maximum file size in checkpoint
-  chunk_byte_size = _DEFAULT_OCDBT_TARGET_DATA_FILE_SIZE
+  chunk_byte_size = DEFAULT_OCDBT_TARGET_DATA_FILE_SIZE
```
Comment (A9isha) [inline MaxText/train.py:150]: Curious why DEFAULT_OCDBT_TARGET_DATA_FILE_SIZE is not used in the other trainers - I think it is because the other trainers have never been used with large enough models and they should also need this logic, so this chunking during checkpoint logic should also be taken out of train.py - but this should be done in a different PR

File: MaxText/sft_trainer.py [hunk 4]
```diff
@@ -143,18 +131,7 @@ def train_loop(config, recorder, state=None):
   last_profiling_step = prof.finished_initial_profile_step
 
   example_batch = None
-  last_step_completion = datetime.datetime.now()
```
Comment (A9isha) [inline MaxText/sft_trainer.py:142]: But previously `last_step_completion` was calculated here after `write_setup_info_to_tensorboard ` work is done
am I reading this right?

in fact I think it should be calculated a bit later right before train step begins

File: MaxText/train.py (hunk 3 above)
Comment (SurbhiJainUSC) [reply MaxText/train.py:150]: `DEFAULT_OCDBT_TARGET_DATA_FILE_SIZE` is used by `save_checkpoint()` in train.py and al other trainers are importing `save_checkpoint()` from train.

File: MaxText/metric_logger.py (hunk 1 above)
Comment (SurbhiJainUSC) [reply MaxText/metric_logger.py:118]: It is used by `write_metrics_to_gcs()`.

File: MaxText/sft_trainer.py (hunk 4 above)
Comment (SurbhiJainUSC) [reply MaxText/sft_trainer.py:142]: It is used to calculate `step_time_delta`.
step_time_delta for step=n is `(time when train step n is completed) - (time when train step n-1 is completed)`

For step>1, calculating `last_step_completion` in `metric_logger.record_train_metrics()` makes sense because it is called just after training step is completed. However, it is unclear to me what should be the right spot for calculating `last_step_completion` for first step. Also, how do we interpret `step_time_delta` for step=0? What is your opinion @A9isha ?

File: MaxText/train.py (hunk 2 above)
Comment (SurbhiJainUSC) [reply MaxText/train.py:786]: Yes, I have verified the logs are same on Tensoboard before and after refactoring.

File: MaxText/sft_trainer.py (hunk 4 above)
Comment (SurbhiJainUSC) [reply MaxText/sft_trainer.py:142]: After looking into it more, I have moved the calculation for `last_step_completion` for the first step to just before we start the training run. This aligns with what you were suggesting previously.

File: MaxText/metric_logger.py (hunk 1 above)
Comment (A9isha) [reply MaxText/metric_logger.py:118]: ah yes - thanks!

File: MaxText/train.py (hunk 3 above)
Comment (A9isha) [reply MaxText/train.py:150]: I see - yeah this functionality needs to be taken out of train.py. It's quite convoluted this way :D

File: MaxText/sft_trainer.py (hunk 4 above)
Comment (SurbhiJainUSC) [reply MaxText/sft_trainer.py:146]: Here are the training logs with and without refactoring for sanity check: gpaste/5145022758125568

**Review by richjames0 — APPROVED** [review]
lgtm

**Review by A9isha — APPROVED** [review]
Thanks Surbhi

**Comment by gobbleturk** [comment]
For reference this undoes the double buffer metric spirit of https://github.com/AI-Hypercomputer/maxtext/pull/390 - we are opting back into simplicity over a small perf gain
